utilities/infra.py

```diff
@@ -275,30 +275,30 @@ def wait_for_pods_deletion(pods):
 
 def get_pod_container_error_status(pod):
     pod_instance_status = pod.instance.status
-    # Check the containerStatuses and if any containers is in waiting state, return that information:
 
+    # Check the containerStatuses and if any containers is in waiting state, return that information:
     for container_status in pod_instance_status.get("containerStatuses", []):
         waiting_container = container_status.get("state", {}).get("waiting")
         if waiting_container:
             return waiting_container["reason"] if waiting_container.get("reason") else waiting_container
 
 
-def get_not_running_pods(pods, filter_pods_by_name=None):
+def get_not_running_pods(pods: list[Pod], filter_pods_by_name: str = "") -> list[dict[str, str]]:
     pods_not_running = []
     for pod in pods:
-        pod_instance = pod.instance
         if filter_pods_by_name and filter_pods_by_name in pod.name:
             LOGGER.warning(f"Ignoring pod: {pod.name} for pod state validations.")
             continue
-        container_status_error = get_pod_container_error_status(pod=pod)
-        if container_status_error:
-            pods_not_running.append({pod.name: container_status_error})
         try:
+            pod_instance = pod.instance
+            if container_status_error := get_pod_container_error_status(pod=pod):
+                pods_not_running.append({pod.name: container_status_error})
+
             # Waits for all pods in a given namespace to be in final healthy state(running/completed).
             # We also need to keep track of pods marked for deletion as not running. This would ensure any
             # pod that was spinned up in place of pod marked for deletion, reaches healthy state before end
             # of this check
-            if pod_instance.metadata.get("deletionTimestamp") or pod_instance.status.phase not in (
+            elif pod_instance.metadata.get("deletionTimestamp") or pod_instance.status.phase not in (
```

Comment:
i think it would make sense to reverse the conditions - i.e first check pod's metadata and phase and only if does not meet that condition iterate over the containers.